IsNegativeDivergentRule.java

```diff
@@ -0,0 +1,111 @@
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+package org.ta4j.core.trading.rules;
+
+import org.ta4j.core.Decimal;
+import org.ta4j.core.Indicator;
+import org.ta4j.core.Rule;
+import org.ta4j.core.TimeSeries;
+import org.ta4j.core.TradingRecord;
+import org.ta4j.core.indicators.helpers.MaxPriceIndicator;
+import org.ta4j.core.trading.rules.AbstractRule;
+
+/**
+ * Indicator-negative-divergent rule ("bearish divergence").
+ * <p>
+ * Satisfied when the values of the {@link Indicator indicator} strict decrease
+ * or is the lowest or is not the highest within the timeFrame and the values of
+ * the other {@link Indicator indicator} strict increase or is the highest or is
+ * not the lowest within a timeFrame. It can test both, strict and non-strict
+ * divergence.
+ */
+public class IsNegativeDivergentRule extends AbstractRule {
+
+	/** The actual indicator */
+	private Indicator<Decimal> ref;
+	/** The other indicator */
+	private Indicator<Decimal> other;
+	/** The timeFrame */
+	private int timeFrame;
+	/** Test for strict divergence. */
+	private boolean useStrictDivergence;
+
+	/**
+	 * Constructor. <br/>
+	 * 
+	 * If "useStrictDivergence" is enabled, divergence is satisfied if values of
+	 * "ref" strict decrease while the values of the "other" indicator strict
+	 * increase.
+	 * 
+	 * @param ref
+	 * @param other
+	 * @param timeFrame
+	 * @param useStrictDivergence
+	 */
+	public IsNegativeDivergentRule(Indicator<Decimal> ref, Indicator<Decimal> other, int timeFrame,
+			boolean useStrictDivergence) {
+		this.ref = ref;
+		this.other = other;
+		this.timeFrame = timeFrame;
+		this.useStrictDivergence = useStrictDivergence;
+	}
+	
+	/**
+	 * Constructor.
+	 * 
+	 * Rule for negative divergence between ref and MaxPriceIndicator.
+	 * 
+	 * @param series
+	 * @param ref
+	 * @param timeFrame
+	 */
+	public IsNegativeDivergentRule(TimeSeries series, Indicator<Decimal> ref, int timeFrame) {
+		this(ref, new MaxPriceIndicator(series), timeFrame, false);
+	}
+
+	@Override
+	public boolean isSatisfied(int index, TradingRecord tradingRecord) {
+		Rule refIsFalling = new IsFallingRule(ref, timeFrame);
+		Rule otherIsRising = new IsRisingRule(other, timeFrame);
+
+		Rule isNegativeDivergent;
+
+		if (useStrictDivergence) {
+			isNegativeDivergent = refIsFalling.and(otherIsRising);
+		} else {
+
+			Rule refIsLowest = new IsLowestRule(ref, timeFrame);
+			Rule otherIsHighest = new IsHighestRule(other, timeFrame);
+			
+			Rule refIsNotHighest = new IsHighestRule(ref, timeFrame).negation();
+			Rule otherIsNotLowest = new IsLowestRule(other, timeFrame).negation();
+
+			isNegativeDivergent = (refIsFalling.or(refIsLowest).or(refIsNotHighest))
+					.and(otherIsRising.or(otherIsHighest).or(otherIsNotLowest));
+		}
+
+		final boolean satisfied = isNegativeDivergent.isSatisfied(index, tradingRecord);
```

Comment:
`index `paramter is just needed here. Maybe you could make case differentiation and initialization of `isNegativeDivergent ` rule in the constructor?

Reply:
I dont understand what you mean. "index" is already provided within the method "isSatisfied()" as is. What is missing? 

Please explain by improving the code.

Reply:
I think it could be stuctured as followed:

´´´
public class IsNegativeDivergentRule extends AbstractRule {


	private final Rule isNegativeDivergent;

	/**
	 * Constructor. <br/>
	 * 
	 * If "useStrictDivergence" is enabled, divergence is satisfied if values of
	 * "ref" strict decrease while the values of the "other" indicator strict
	 * increase.
	 * 
	 * @param ref the first indicator
	 * @param other the other indicator
	 * @param timeFrame the time frame
	 * @param useStrictDivergence true if only strict dicrease 
	 */
	public IsNegativeDivergentRule(Indicator<Decimal> ref, Indicator<Decimal> other, int timeFrame,
			boolean useStrictDivergence) {
		Indicator<Decimal> ref = ref; // can be local
		Indicator<Decimal> other = other; // can be local
		int timeFrame = timeFrame; // can be local
		this.useStrictDivergence = useStrictDivergence;
		
		// init rule isNegativeDivergent here...
		Rule refIsFalling = new IsFallingRule(ref, timeFrame);
		Rule otherIsRising = new IsRisingRule(other, timeFrame);
		
		if (useStrictDivergence) {
			isNegativeDivergent = refIsFalling.and(otherIsRising);
		} else {
			Rule refIsLowest = new IsLowestRule(ref, timeFrame);
			Rule otherIsHighest = new IsHighestRule(other, timeFrame);
			
			Rule refIsNotHighest = new IsHighestRule(ref, timeFrame).negation();
			Rule otherIsNotLowest = new IsLowestRule(other, timeFrame).negation();

			isNegativeDivergent = (refIsFalling.or(refIsLowest).or(refIsNotHighest))
					.and(otherIsRising.or(otherIsHighest).or(otherIsNotLowest));
		}
	}
	
	/**
	 * Constructor.
	 * 
	 * Rule for negative divergence between ref and MaxPriceIndicator.
	 * 
	 * @param series
	 * @param ref
	 * @param timeFrame
	 */
	public IsNegativeDivergentRule(TimeSeries series, Indicator<Decimal> ref, int timeFrame) {
		this(ref, new MaxPriceIndicator(series), timeFrame, false);
	}

	@Override
	public boolean isSatisfied(int index, TradingRecord tradingRecord) {
		
		// ...just check isSatisfied here
		final boolean satisfied = isNegativeDivergent.isSatisfied(index, tradingRecord);
		traceIsSatisfied(index, satisfied);
		return satisfied;
	}
}
´´´

Reply:
What is the benefit of moving the logic within the constructor?

Please look at "IsFallingRule.java", "IsRisingRule.java", I had also put the logic within the isSatisfied()-method. Should I change these files and move the logic within the constructor, too?